## Homework

Q1. Pandas version
What's the version of Pandas that you installed?

In [2]:
import pandas as pd
import numpy as np

In [3]:
pd.__version__

'2.3.1'

### Getting the data

In [4]:
df = pd.read_csv("https://raw.githubusercontent.com/alexeygrigorev/datasets/master/car_fuel_efficiency.csv")

In [5]:
df.head()

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
0                  170            3.0       159.0     3413.433759   
1                  130            5.0        97.0     3149.664934   
2                  170            NaN        78.0     3079.038997   
3                  220            4.0         NaN     2542.392402   
4                  210            1.0       140.0     3460.870990   

   acceleration  model_year  origin fuel_type         drivetrain  num_doors  \
0          17.7        2003  Europe  Gasoline    All-wheel drive        0.0   
1          17.8        2007     USA  Gasoline  Front-wheel drive        0.0   
2          15.1        2018  Europe  Gasoline  Front-wheel drive        0.0   
3          20.2        2009     USA    Diesel    All-wheel drive        2.0   
4          14.4        2009  Europe  Gasoline    All-wheel drive        2.0   

   fuel_efficiency_mpg  
0            13.231729  
1            13.688217  
2            14.246341  
3            16.912736  
4            12.488369

Q2. Records count

How many records are in the dataset?

In [6]:
# How many records are in the dataset? (records, columns)
df.shape

(9704, 11)

Answer: 9704 records

Q3. Fuel types

How many fuel types are presented in the dataset?

In [7]:
# How many fuel types are presented in the dataset?
df['fuel_type'].nunique()

2

Answer: 2

Q4. Missing values

How many columns in the dataset have missing values?

In [8]:
# explore columns in the dataframe to know which ones have missing values
df.isna().any()

engine_displacement    False
num_cylinders           True
horsepower              True
vehicle_weight         False
acceleration            True
model_year             False
origin                 False
fuel_type              False
drivetrain             False
num_doors               True
fuel_efficiency_mpg    False
dtype: bool

In [9]:
# number of missing values in each of the columns
df.isna().sum()

engine_displacement      0
num_cylinders          482
horsepower             708
vehicle_weight           0
acceleration           930
model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors              502
fuel_efficiency_mpg      0
dtype: int64

In [10]:
# How many columns in the dataset have missing values?
print(df.isna().any().sum())

4


Answer: 4

Q5. Max fuel efficiency

What's the maximum fuel efficiency of cars from Asia?

In [11]:
# 1️⃣  Filter the records where the origin is equal to Asia
asia_cars = df[df['origin'] == 'Asia']

# 2️⃣  Select only the mpg column from those rows
asia_fuel_efficiency_mpg = asia_cars['fuel_efficiency_mpg']

# 3️⃣  Find the maximum fuel efficiency (mpg) value
max_fuel_efficiency_mpg = asia_fuel_efficiency_mpg.max()

# 4️⃣  Display the result
print(max_fuel_efficiency_mpg)


23.759122836520497


In [12]:
# What's the maximum fuel efficiency of cars from Asia?
print(round(df.loc[df['origin'] == 'Asia', 'fuel_efficiency_mpg'].max(), 2))

23.76


Answer: 23.75

Q6. Median value of horsepower
1. Find the median value of horsepower column in the dataset.
2. Next, calculate the most frequent value of the same horsepower column.
3. Use fillna method to fill the missing values in horsepower column with the most frequent value from the previous step.
4. Now, calculate the median value of horsepower once again.

### Has it changed?

- Yes, it increased
- Yes, it decreased
- No

In [20]:
# 1. Find the median value of horsepower column in the dataset.
print(df['horsepower'].median())

152.0


In [21]:
# 2. Next, calculate the most frequent value of the same horsepower column.
most_freq_hp_value = df['horsepower'].mode()[0]
print(most_freq_hp_value)

152.0


In [22]:
# 3. Use fillna method to fill the missing values in horsepower column with the most frequent value from the previous step.
df['horsepower'] = df['horsepower'].fillna(most_freq_hp_value)

In [23]:
# 4. Now, calculate the median value of horsepower once again.
print(df['horsepower'].median())

152.0


Answer: Yes, it increased

Q7. Sum of weights
1. Select all the cars from Asia
2. Select only columns vehicle_weight and model_year
3. Select the first 7 values
4. Get the underlying NumPy array. Let's call it X.
5. Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
6. Invert XTX.
7. Create an array y with values [1100, 1300, 800, 900, 1000, 1100, 1200].
8. Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
9. What's the sum of all the elements of the result?

_Note: You just implemented linear regression. We'll talk about it in the next lesson._

- 0.051
- 0.51
- 5.1
- 51

In [25]:
# 1️ -  Select all cars from Asia
cars_from_asia = df[df['origin'] == 'Asia']

# 2️.  Select only columns vehicle_weight and model_year
cars_from_asia_subset = cars_from_asia[['vehicle_weight', 'model_year']]

# 3️.  Select the first 7 values
cars_from_asia_7 = cars_from_asia_subset.head(7)

# 4️.  Get the underlying NumPy array. Let's call it X.
X = cars_from_asia_7.to_numpy()

# 5️.  Compute matrix-matrix multiplication between the transpose of X and X. 
# To get the transpose, use X.T. Let's call the result XTX.
XTX = X.T.dot(X)

# 6️.  Invert XTX.
XTX_inv = np.linalg.inv(XTX)

# 7️.  Create an array y with values [1100, 1300, 800, 900, 1000, 1100, 1200].
y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])

# 8️.  Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
# e.g. w = (XTX)⁻¹ Xᵀ y
w = XTX_inv.dot(X.T).dot(y)

# 9️.  What's the sum of all the elements of the result?
result_sum = w.sum()
print(result_sum)

0.5187709081074016
